### PR TITLE
grub2: fix cross-compilation

### PR DESCRIPTION
**Pull request by r-burns**
* gettext is used for native binaries, not hosted libs
* grub2 expects `BUILD_CC` to point to a native C compiler
* add `strictDeps = true` to prevent regressions

###### Motivation for this change

Fixes e.g. `pkgsCross.aarch64-multiplatform.grub2`

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (nixos/tests/grub.nix)
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112857` at 00f550fc run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>grub2_pvgrub_image</li>
    <li><a href="https://gist.github.com/89ce33098b56ae0df4321b4099906611">grub2_xen</a></li></li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/465118b7ae22c370d77a527e1b26b036">grub2</a></li> (grub2_full ,grub2_light)</li>
    <li><a href="https://gist.github.com/84c6a83ffb95f2db043a650436408b18">grub2_efi</a></li></li>
    <li><a href="https://gist.github.com/392f3ccc0edbe7265601d54ca970a319">woeusb</a></li></li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/tools/misc/grub/2.0x.nix:116:3:
```
    |
116 |   meta = with lib; {
    |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/tools/misc/grub/2.0x.nix:54:5:
```
   |
54 |     ./fix-bash-completion.patch
   |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `unifont`.
Near pkgs/tools/misc/grub/2.0x.nix:2:62:
```
  |
2 | , gettext, ncurses, libusb-compat-0_1, freetype, qemu, lvm2, unifont, pkg-config
  |                                                              ^
```
</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112857` at 00f550fc run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixosTest-test</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/3f219b0cffe3918e0f94ca6430e4c9da">grub2</a></li> (grub2_full ,grub2_light)</li>
    <li><a href="https://gist.github.com/e0c8c6238f466ece4e26437fdc2c2351">grub2_efi</a></li></li>
    <li>grub2_pvgrub_image</li>
    <li><a href="https://gist.github.com/7d944517b29d3bcf08ecddc252a85932">grub2_xen</a></li></li>
    <li><a href="https://gist.github.com/bb1bf6fe16a7a96b31fef8617839e2df">os-prober</a></li></li>
    <li><a href="https://gist.github.com/7d51ced7627c62739451796e6f36c88c">woeusb</a></li></li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/tools/misc/grub/2.0x.nix:116:3:
```
    |
116 |   meta = with lib; {
    |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/tools/misc/grub/2.0x.nix:54:5:
```
   |
54 |     ./fix-bash-completion.patch
   |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `unifont`.
Near pkgs/tools/misc/grub/2.0x.nix:2:62:
```
  |
2 | , gettext, ncurses, libusb-compat-0_1, freetype, qemu, lvm2, unifont, pkg-config
  |                                                              ^
```
</li>
  </ul>
</details>

**Comment by symphorien** [comment]
Maybe add `strictDeps=true` to avoid potential future regressions.

**Comment by r-burns** [comment]
Good idea, done.

**Comment by symphorien** [comment]
@GrahamcOfBorg test installer.simple installer.grub1 installer.simpleUefiGrub

**Comment by symphorien** [comment]
(if you feel like it, maybe add a few tests to passthru.test *en passant*, but I'll merge even either way)

**Comment by symphorien** [comment]
Spurious ofborg failure
```
machine # warning: unable to download 'https://ftpmirror.gnu.org/binutils/binutils-2.35.1.tar.bz2': Couldn't resolve host name (6); retrying in 2712 ms
machine # error: unable to download 'https://salsa.debian.org/debian/keyutils/raw/4cecffcb8e2a2aa4ef41777ed40e4e4bcfb2e5bf/debian/patches/Make-build-reproducible.patch': Couldn't resolve host name (6)
machine # builder for '/nix/store/f450ks5xaq790h43nc5r78x69cr4721q-Make-build-reproducible.patch.drv' failed with exit code 1
```
Let's try if we can restart it
@GrahamcOfBorg test installer.simple installer.grub1 installer.simpleUefiGrub

**Comment by r-burns** [comment]
Hm, weird. Let me give it a try on my rpi4.

**Comment by r-burns** [comment]
Ah, I think some of these tests are already failing on master. Not sure what should be done in that case.

**Comment by r-burns** [comment]
I don't think I've ever used passthru.tests before but it seems like a good thing to do going forward. Let me know if what I added is similar to what you had in mind.

**Comment by symphorien** [comment]
Since grub is kind of critical, let's just wait for the tests to be fixed on master.

**Comment by symphorien** [comment]
(Actually the test failure is not spurious: it's the nixos test vm which fails to download some sources when it runs nixos-install without network).

File: pkgs/tools/misc/grub/2.0x.nix [hunk 1]
```diff
@@ -88,6 +92,8 @@ stdenv.mkDerivation rec {
       ./bootstrap --no-git --gnulib-srcdir=${gnulib}
 
       substituteInPlace ./configure --replace '/usr/share/fonts/unifont' '${unifont}/share/fonts'
+
+      export BUILD_CC=${pkgsBuildBuild.stdenv.cc}/bin/cc
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/grub/2.0x.nix:96]: I would prefer if this would be moved to configureFlags.

File: pkgs/tools/misc/grub/2.0x.nix [hunk 2]
```diff
@@ -110,6 +116,14 @@ stdenv.mkDerivation rec {
     sed -i $out/lib/grub/*/modinfo.sh -e "/grub_target_cppflags=/ s|'.*'|' '|"
   '';
 
+  passthru.tests = {
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/grub/2.0x.nix:119]: Can we move the tests into another PR? I would like to have cross compiling support and they block that.

Comment (r-burns) [reply pkgs/tools/misc/grub/2.0x.nix:119]: Ok I can remove this. Can you explain why these would block cross-compiling? Obviously my whole motivation here is cross-compilation so I don't want to harm that :)

Comment (symphorien) [reply pkgs/tools/misc/grub/2.0x.nix:119]: I disagree. Testing goes before cross compiling. Especially for grub, and especially when a fix is provided in https://github.com/NixOS/nixpkgs/pull/112961

Comment (SuperSandro2000) [reply pkgs/tools/misc/grub/2.0x.nix:119]: > Can you explain why these would block cross-compiling?

Because they are failing.

> Especially for grub

also true.

**Comment by symphorien** [comment]
@GrahamcOfBorg eval

**Comment by symphorien** [comment]
thanks
